wagtail/models/__init__.py

```diff
@@ -979,6 +979,33 @@ class Page(AbstractPage, index.Indexed, ClusterableModel, metaclass=PageBase):
     promote_panels = []
     settings_panels = []
 
+    @staticmethod
+    def route_for_request(request, path):
+        """
+        Find the page object for this HTTP request object. The page, args, and
+        kwargs will be cached via request._wagtail_route_for_request
+        """
+        if not hasattr(request, '_wagtail_route_for_request'):
+            try:
+                # we need a valid Site object corresponding to this request in order to proceed
+                site = Site.find_for_request(request)
+                if not site:
+                    raise Http404
+    
+                path_components = [component for component in path.split("/") if component]
+                request._wagtail_route_for_request = site.root_page.localized.specific.route(
+                    request, path_components
+                )
+            except Http404:
+                request._wagtail_route_for_request = None
+        return request._wagtail_route_for_request
+    
+    @staticmethod
+    def find_for_request(*args):
+        result = Page.route_for_request(*args)
```

Comment:
Nit: I suggest we be explicit on arguments here
```suggestion
    def find_for_request(request, path):
        result = Page.route_for_request(request, path)
```